### PR TITLE
Lightning faucet

### LINKED ISSUE #145: Create lightning faucet
Our wallet onboarding path expects the user to transfer funds from another lightning wallet. However this won't be feasible for our tests (regtest, signet or testnet). Consequently we need some kind of lightning faucet that has an open channel with our coordinator.

The flow could be the following.

1. Tester creates an invoice in the app
2. Tester copies the decoded invoice string and navigates his browser to our public lightning faucet.
3. Tester enters pastes the invoice to the faucet and submits it.
4. The faucet automatically pays the invoice and the tester gets an channel with the coordinator.

### PR DESCRIPTION
**Pull request by holzeis**
This PR introduces a lightning faucet based on lnd. Thus addressing two topics at once.

1. How to onboard a 10101 mobile user.
2. Are we compatible with lnd.

The configuration is pretty messy and I am still not 100% sure if I understand all values. But after a hell lot of debugging and trial and error we finally found a setup that was working with lnd. Some more investigation is required here though.

TODOs, that will be addressed in follow up PRs.

- [x] Add automated tests to ensure we stay compatible with lnd - onboard test case #170 
- [ ] Get rid of chopsticks - we don't need that service. the faucet is exposing everything we need. #171
- [ ] The macaroons.db is changed upon every docker-compose up. This is not critical as you can ignore this changes, but it's cumbersome to get a dirty workspace. Note, it also doesn't matter if you commit the changes. (We need this committed, as the macaroons.db contains the secret keys that match the admin.macaroon used in the faucet proxy).

resolves #145

### CODE REVIEW

File: crates/ln-dlc-node/src/node.rs [hunk 1]
```diff
@@ -798,10 +773,17 @@ pub(crate) fn app_config() -> UserConfig {
         },
         channel_handshake_limits: ChannelHandshakeLimits {
             max_minimum_depth: 1,
+            trust_own_funding_0conf: true,
```
Comment (da-kami) [inline crates/ln-dlc-node/src/node.rs:776]: 🙃 Maybe a comment for this option would also be useful

**Review by da-kami — COMMENTED** [review]
I had a quick scan over this and did not find any issues; I'll see to have another look on Monday and also give it a try on my machine (have not so far)!

File: crates/ln-dlc-node/src/node.rs [hunk 2]
```diff
@@ -789,7 +764,7 @@ pub(crate) fn app_config() -> UserConfig {
         channel_handshake_config: ChannelHandshakeConfig {
             // this is needed as otherwise the config between the coordinator and us diverges and we
             // can't open channels.
-            announced_channel: true,
+            announced_channel: false,
```
[the remark this answers is not in the document]
Comment (holzeis) [reply crates/ln-dlc-node/src/node.rs:768]: I've updated the comment.

File: crates/ln-dlc-node/src/node.rs (hunk 1 above)
Comment (holzeis) [reply crates/ln-dlc-node/src/node.rs:781]: well, this is actually just the default config and does not change anything. 🙈

File: crates/ln-dlc-node/src/tests/just_in_time_channel.rs [hunk 3]
```diff
@@ -26,9 +26,13 @@ async fn just_in_time_channel() {
...
-        .open_channel(coordinator.info, payer_outbound_liquidity_sat, 0)
+    let payer_outbound_liquidity_sat = 25_000;
+    let payer_coordinator_channel_details = coordinator
+        .open_channel(
+            &payer,
+            payer_outbound_liquidity_sat,
+            payer_outbound_liquidity_sat,
+        )
         .await
         .unwrap();
```
Comment (luckysori) [inline crates/ln-dlc-node/src/tests/just_in_time_channel.rs:37]: :wrench: I think it's clearer if we only fund one side (the payer). Although it might not be possible given that we let the coordinator initiate the channel creation. At the very least, we should not fund the coordinator with the `payer_outbound_liquidity_sat` variable.

File: mobile/pubspec.lock [hunk 4]
```diff
@@ -950,5 +950,5 @@ packages:
     source: hosted
     version: "2.0.3"
 sdks:
-  dart: ">=2.19.0 <3.0.0"
+  dart: ">=2.19.0 <4.0.0"
```
Comment (luckysori) [inline mobile/pubspec.lock:953]: :question: Any idea why this changed?

File: crates/ln-dlc-node/src/tests/multi_hop_payment.rs [hunk 5]
```diff
@@ -17,21 +16,18 @@ async fn multi_hop_payment() {
...
-    router.open_channel(payee.info, 20_000, 0).await.unwrap();
+    router.open_channel(&payer, 20_000, 20_000).await.unwrap();
+    router.open_channel(&payee, 20_000, 20_000).await.unwrap();
 
     let payer_balance_before = payer.get_ldk_balance();
     let router_balance_before = router.get_ldk_balance();
     let payee_balance_before = payee.get_ldk_balance();
 
-    // Add 6 confirmations for the channels to be announced
-    mine(6).await;
```
Comment (luckysori) [inline crates/ln-dlc-node/src/tests/multi_hop_payment.rs:33]: :wrench: We should remove the subsequent syncs and sleep since we aren't mining anything.

File: crates/ln-dlc-node/src/tests/mod.rs [hunk 6]
```diff
@@ -207,6 +238,7 @@ async fn fund_and_mine(address: bitcoin::Address, amount: bitcoin::Amount) {
 }
 
 /// Instructs `nigiri-chopsticks` to mine a block.
+#[allow(dead_code)]
```
Comment (luckysori) [inline crates/ln-dlc-node/src/tests/mod.rs:241]: :wrench: We should delete this since we don't use `chopsticks` anymore.

**Review by luckysori — APPROVED** [review]
Thank you for working on this! We learnt a lot along the way :tada:

File: .github/workflows/tests-ln-dlc-node.yml [hunk 7]
```diff
@@ -41,7 +41,7 @@ jobs:
       - uses: Swatinem/rust-cache@v2.2.0
       - name: Start containers
         run: |
-          docker-compose up -d
+          docker-compose up -d chopsticks
```
[the remark this answers is not in the document]
Comment (holzeis) [reply .github/workflows/tests-ln-dlc-node.yml:44]: Because we don't need to start esplora for the tests. Saves little bit of resources 😉

File: crates/ln-dlc-node/src/tests/mod.rs (hunk 6 above)
Comment (holzeis) [reply crates/ln-dlc-node/src/tests/mod.rs:241]: We still do. I created a follow up ticket for removing chopsticks. In the meantime I figured it would be best to keep that function.

File: mobile/pubspec.lock (hunk 4 above)
Comment (holzeis) [reply mobile/pubspec.lock:953]: Nope!

File: coordinator/src/routes.rs [hunk 8]
```diff
@@ -29,7 +29,7 @@ pub async fn post_fake_scid(
 pub async fn get_new_address(node: &State<Arc<Node>>) -> Result<Json<String>, HttpApiProblem> {
     let address = node.wallet.get_new_address().map_err(|e| {
         HttpApiProblem::new(StatusCode::INTERNAL_SERVER_ERROR)
-            .title("Invalid public key")
+            .title("Failed to get new address")
             .detail(format!("Failed to get new address: {e:#}"))
```
[the remark this answers is not in the document]
Comment (holzeis) [reply coordinator/src/routes.rs]: https://github.com/get10101/10101/pull/169/commits/fb7fffde10772db8785d95d184acc37fa4e68ae8

File: crates/ln-dlc-node/src/node.rs [hunk 9]
```diff
@@ -423,7 +424,9 @@ impl Node {
                     .expect("Failed to bind to listen port");
                 loop {
                     let peer_manager = peer_manager.clone();
-                    let (tcp_stream, _) = listener.accept().await.unwrap();
+                    let (tcp_stream, addr) = listener.accept().await.unwrap();
+
+                    tracing::info!(%addr, "Received inbound connection");
```
[the remark this answers is not in the document]
Comment (holzeis) [reply crates/ln-dlc-node/src/node.rs]: https://github.com/get10101/10101/pull/169/commits/91640cbef71a297e0ca8da444fa1f7401e2924a7

File: crates/ln-dlc-node/src/node.rs [hunk 10]
```diff
@@ -787,21 +762,33 @@ impl Node {
...
         channel_handshake_config: ChannelHandshakeConfig {
-            // this is needed as otherwise the config between the coordinator and us diverges and we
-            // can't open channels.
-            announced_channel: true,
+            // The app will only accept private channels. As we are forcing the apps announced
+            // channel preferences, the coordinator needs to override this config to match the apps
+            // preferences.
+            announced_channel: false,
+            // The minimum amount of confirmations before the inbound channel is deemed useable,
+            // between the counterparties
```
[the remark this answers is not in the document]
Comment (holzeis) [reply crates/ln-dlc-node/src/node.rs]: https://github.com/get10101/10101/pull/169/commits/91640cbef71a297e0ca8da444fa1f7401e2924a7

File: crates/ln-dlc-node/src/ln/event_handler.rs [hunk 11]
```diff
@@ -440,12 +442,24 @@ impl EventHandler {
...
+                    expected_outbound_amount_msat * 2 / 1000,
+                );
+
+                let mut user_config = coordinator_config();
+                // We are overwriting the coordinators channel handshake configuration to prevent
+                // the just-in-time-channel from being announced (private). This is required as both
+                // parties need to agree on this configuration. For other channels, like with the
+                // channel to the faucet we want this channel to be announced (public).
+                // NOTE: we want private channels with the mobile app, as this will allow us to make
+                // use of 0-conf channels.
```
[the remark this answers is not in the document]
Comment (holzeis) [reply crates/ln-dlc-node/src/ln/event_handler.rs]: https://github.com/get10101/10101/pull/169/commits/91640cbef71a297e0ca8da444fa1f7401e2924a7

File: crates/ln-dlc-node/src/tests/add_dlc.rs [hunk 12]
```diff
@@ -23,18 +23,24 @@ async fn given_lightning_channel_then_can_add_dlc_channel() {
...
+        .fund(Amount::from_btc(0.1).unwrap())
         .await
         .unwrap();
 
+    coordinator.open_channel(&app, 50000, 50000).await.unwrap();
+    let channel_details = app.channel_manager.list_usable_channels();
+    let channel_details = channel_details
+        .iter()
+        .find(|c| c.counterparty.node_id == coordinator.info.pubkey)
+        .unwrap();
```
[the remark this answers is not in the document]
Comment (holzeis) [reply crates/ln-dlc-node/src/tests/add_dlc.rs]: https://github.com/get10101/10101/pull/169/commits/dc8c5e2b43923035927af3e2710f5e9e821052c9

File: crates/ln-dlc-node/src/tests/just_in_time_channel.rs (hunk 3 above)
Comment (holzeis) [reply crates/ln-dlc-node/src/tests/just_in_time_channel.rs]: https://github.com/get10101/10101/pull/169/commits/dc8c5e2b43923035927af3e2710f5e9e821052c9

File: crates/ln-dlc-node/src/tests/multi_hop_payment.rs (hunk 5 above)
Comment (holzeis) [reply crates/ln-dlc-node/src/tests/multi_hop_payment.rs:33]: the syncs are still needed to get every node up to date on the channels, but I removed the sleep.

https://github.com/get10101/10101/pull/169/commits/dc8c5e2b43923035927af3e2710f5e9e821052c9

File: crates/ln-dlc-node/src/tests/add_dlc.rs (hunk 12 above)
[the remark this answers is not in the document]
Comment (holzeis) [reply crates/ln-dlc-node/src/tests/add_dlc.rs]: Since I turned the initiator of the channel around I can not use the channel_details returned by the coordinator. Hence I am looking for the channel details in the app channel_manager.

File: docker-compose.yml [hunk 13]
```diff
@@ -90,6 +92,40 @@ services:
...
+  lnd:
+    container_name: lnd
+    image: lightninglabs/lnd:v0.15.5-beta
+    depends_on:
+      - bitcoin
+    command:
+      - --configfile=/var/lib/lnd/lnd.conf
+    volumes:
+      - ./faucet/lnd.conf:/var/lib/lnd/lnd.conf
+      - ./faucet/macaroons.db:/root/.lnd/data/chain/bitcoin/regtest/macaroons.db
```
Comment (bonomat) [inline docker-compose.yml:104]: This will create a directory locally. Is this needed or can we use a virtual folder?

File: docker-compose.yml [hunk 14]
```diff
@@ -90,6 +92,40 @@ services:
       - 3000:3000
     restart: unless-stopped
 
+  lnd:
+    container_name: lnd
+    image: lightninglabs/lnd:v0.15.5-beta
+    depends_on:
+      - bitcoin
+    command:
+      - --configfile=/var/lib/lnd/lnd.conf
```
Comment (bonomat) [inline docker-compose.yml:101]: Can we pass in the related conf via arg so that we don't have to deal with local filestorage?

File: docker-compose.yml [hunk 15]
```diff
@@ -90,6 +92,40 @@ services:
...
+
+  faucet:
+    image: nginx
+    container_name: faucet
+    ports:
+      - 8080:80
+    volumes:
+      - ./faucet/faucet.html:/usr/share/nginx/html/faucet/index.html
+      - ./faucet/open_channel.html:/usr/share/nginx/html/channel/index.html
+      - ./faucet/default.conf:/etc/nginx/conf.d/default.conf
```
Comment (bonomat) [inline docker-compose.yml:123]: I guess my two comments above are obsolete because we still need these files :D 🙈

File: faucet/lnd.conf [hunk 16]
```diff
@@ -0,0 +1,86 @@
...
+autopilot.active=false
+
+[tor]
+tor.active=false
+
+[db]
+db.backend=bolt
+
+[bolt]
+db.bolt.auto-compact=true
```
Comment (bonomat) [inline faucet/lnd.conf:86]: 🙃
Missing newline

File: faucet/open_channel.html [hunk 17]
```diff
@@ -0,0 +1,125 @@
+<!DOCTYPE html>
```
Comment (bonomat) [inline faucet/open_channel.html:1]: this comment is meant for above for `faucet/macaroons.db`: did you intent to check this in?

File: crates/ln-dlc-node/src/ln/event_handler.rs [hunk 18]
```diff
@@ -440,12 +442,24 @@ impl EventHandler {
...
+                    JUST_IN_TIME_CHANNEL_OUTBOUND_LIQUIDITY_SAT,
+                    expected_outbound_amount_msat * 2 / 1000,
+                );
+
+                let mut user_config = coordinator_config();
+                // We are overwriting the coordinators channel handshake configuration to prevent
+                // the just-in-time-channel from being announced (private). This is required as both
+                // parties need to agree on this configuration. For other channels, like with the
+                // channel to the faucet we want this channel to be announced (public).
+                // NOTE: we want private channels with the mobile app, as this will allow us to make
```
Comment (bonomat) [inline crates/ln-dlc-node/src/ln/event_handler.rs:460]: Are you saying that only private channels can have 0-conf?

File: crates/ln-dlc-node/src/tests/mod.rs [hunk 19]
```diff
@@ -143,12 +143,12 @@ impl Node {
     /// the peers involved!! It may not always work.
     async fn open_channel(
         &self,
-        peer: NodeInfo,
+        peer: &Node,
```
Comment (bonomat) [inline crates/ln-dlc-node/src/tests/mod.rs:146]: Was this change really needed given that you make use of `peer.info` anyways?

File: crates/ln-dlc-node/src/node.rs [hunk 20]
```diff
@@ -573,7 +575,7 @@ impl Node {
                     base_msat: 1000,
                     proportional_millionths: LIQUIDITY_ROUTING_FEE_MILLIONTHS,
                 },
-                cltv_expiry_delta: MIN_CLTV_EXPIRY_DELTA,
+                cltv_expiry_delta: 144,
```
Comment (bonomat) [inline crates/ln-dlc-node/src/node.rs:548]: Can we add a comment why this needs to be `144` please?

File: crates/ln-dlc-node/src/node.rs [hunk 21]
```diff
@@ -564,6 +565,7 @@ impl Node {
             .payment_hash(sha256::Hash::from_slice(&payment_hash.0)?)
             .payment_secret(payment_secret)
             .timestamp(SystemTime::now())
+            .min_final_cltv_expiry(133) // default: 9 todo: check why lnd payments only work with a higher number
```
Comment (bonomat) [inline crates/ln-dlc-node/src/node.rs:538]: I don't understand the beginning of the comment

File: crates/ln-dlc-node/src/node.rs [hunk 22]
```diff
@@ -814,23 +828,45 @@ pub(crate) fn app_config() -> UserConfig {
...
             max_inbound_htlc_value_in_flight_percent_of_channel: 10,
             ..Default::default()
         },
         channel_handshake_limits: ChannelHandshakeLimits {
+            // The minimum amount of confirmations before the outbound channel is deemed useable,
+            // between the counterparties
             max_minimum_depth: 1,
             trust_own_funding_0conf: true,
-            force_announced_channel_preference: false,
+            // Enforces incoming channels to the coordinator to be public! We
```
Comment (bonomat) [inline crates/ln-dlc-node/src/node.rs:817]: Because of this, do we need to ensure that a routing hint is always added to an invoice generated by the app?

File: README.md [hunk 23]
```diff
@@ -104,3 +104,29 @@ or in short
...
+The following command will start a regtest bitcoin node, electrs, chopsticks, esplora, lnd node and a http server allowing you to faucet your bitcoin and lightning wallet.
+
+```bash
+docker-compose up
+```
+
+### How to faucet your lightning wallet.
+
+#### Setup
+
```
Comment (bonomat) [inline README.md:119]: I failed at step 3 otherwise :) 
```suggestion
0. Start the coordinator
```

File: README.md [hunk 24]
```diff
@@ -104,3 +104,29 @@ or in short
...
+1. Open `http://localhost:8080/faucet`
+2. Ensure you have enough balance on your bitcoin wallet. Hit the mine button a couple of times if not.
+3. Get a new address of your coordinator by running `curl http://localhost:8000/api/newaddress`
+4. Faucet some coins to your coordinator wallet. Hit the mine button afterwards so the transaction gets into a block.
+5. Open `http://localhost:8080/channel`
```
Comment (bonomat) [inline README.md:124]: This tries to forward me to `http://localhost:80/channel` which fails because that port is not exported:

```
curl http://localhost:8080/channel -v

*   Trying 127.0.0.1:8080...
* Connected to localhost (127.0.0.1) port 8080 (#0)
> GET /channel HTTP/1.1
> Host: localhost:8080
> User-Agent: curl/7.86.0
> Accept: */*
>
* Mark bundle as not supporting multiuse
< HTTP/1.1 301 Moved Permanently
< Server: nginx/1.23.3
< Date: Sun, 26 Feb 2023 23:42:21 GMT
< Content-Type: text/html
< Content-Length: 169
< Location: http://localhost/channel/
< Connection: keep-alive
<
<html>
<head><title>301 Moved Permanently</title></head>
<body>
<center><h1>301 Moved Permanently</h1></center>
<hr><center>nginx/1.23.3</center>
</body>
</html>
* Connection #0 to host localhost left intact
* ```

Because of this I can't test any further.

**Review by bonomat — CHANGES_REQUESTED** [review]
Unfortunately it does not work for me :/ See comment on the readme.

File: README.md (hunk 24 above)
Comment (luckysori) [reply README.md:124]: I had the same problem using `curl`. But I could get around it using a web browser _if_ I included the trailing forward slash i.e. `http://localhost:8080/channel/`.

Comment (bonomat) [reply README.md:124]: thanks @luckysori : this fixes it.

Comment (bonomat) [inline README.md:124]:
```suggestion
5. Open `http://localhost:8080/channel/` (note: ensure to add the trailing `/` as otherwise nginx will try to redirect the call)
```

**Review by bonomat — APPROVED** [review]

File: crates/ln-dlc-node/src/ln/event_handler.rs (hunk 18 above)
Comment (holzeis) [reply crates/ln-dlc-node/src/ln/event_handler.rs:460]: yes.

File: faucet/open_channel.html (hunk 17 above)
Comment (holzeis) [reply faucet/open_channel.html:1]: yes, as the macaroons.db contains the keys for the macaroon generation. We need to keep them the same so we can pass the admin.macaroon statically in our faucet proxy. Nothing you would consider for production, but more convenient for the tests.

File: crates/ln-dlc-node/src/tests/mod.rs (hunk 19 above)
Comment (holzeis) [reply crates/ln-dlc-node/src/tests/mod.rs:146]: no it wasn't 😅 . I reverted it.

File: crates/ln-dlc-node/src/node.rs (hunk 20 above)
Comment (holzeis) [reply crates/ln-dlc-node/src/node.rs:548]: Actually it is also working with MIN_CLTV_EXPIRY_DELTA (42). We've changed it in some previous attempts as this was the setting used by Breez and Phoenix.

File: crates/ln-dlc-node/src/node.rs (hunk 21 above)
Comment (holzeis) [reply crates/ln-dlc-node/src/node.rs:538]: I've updated the comment. I hope it's clearer now.

File: crates/ln-dlc-node/src/node.rs (hunk 22 above)
Comment (holzeis) [reply crates/ln-dlc-node/src/node.rs:817]: Not sure if I fully understand the context of your question. But yes I think we will always need to add a routing hint to the invoice if we want to send a payment over a private channel.

File: README.md (hunk 24 above)
Comment (holzeis) [reply README.md:124]: I've updated the `README.md`

File: docker-compose.yml (hunk 15 above)
Comment (holzeis) [reply docker-compose.yml:123]: yes, unfortunately.. alternatively we could build a small docker image, that contains these files.

File: docker-compose.yml (hunk 14 above)
Comment (holzeis) [reply docker-compose.yml:101]: It should be possible, but as you've pointed out below, we would still need to deal with local file storage.

File: docker-compose.yml (hunk 13 above)
Comment (holzeis) [reply docker-compose.yml:104]: That we mount the `macaroons.db` like that is not ideal, but I did not find a better way to ensure that the `admin.macaroon` will not change on every fresh start.
